IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,683 @@
+                          error:(NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return NO;
+}
+
+// Read information about the app which most-recently removed a token.
+- (NSDictionary *)wipeInfo:(__unused id<MSIDRequestContext>)context
+                     error:(NSError **)error
+{
+    [self createUnimplementedError:error context:context];
+    return nil;
+}
+
+#pragma mark - clear
+
+// A test-only method that deletes all items from the cache for the given context.
+- (BOOL)clearWithContext:(id<MSIDRequestContext>)context
+                   error:(NSError **)error
+
+{
+    MSID_LOG_WARN(context, @"Clearing the whole context. This should only be executed in tests");
+
+    // Delete all accounts for the keychainGroup
+    NSMutableDictionary *query = [self.defaultAccountQuery mutableCopy];
+    query[(id)kSecMatchLimit] = (id)kSecMatchLimitAll;
+    MSID_LOG_VERBOSE(context, @"Trying to delete keychain items...");
+    OSStatus status = SecItemDelete((CFDictionaryRef)query);
+    MSID_LOG_VERBOSE(context, @"Keychain delete status: %d", (int)status);
+
+    if (status != errSecSuccess && status != errSecItemNotFound)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDKeychainErrorDomain, status, @"Failed to remove items from keychain.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Failed to delete keychain items (status: %d)", (int)status);
+
+        return NO;
+    }
+
+    return YES;
+}
+
+#pragma mark - Utilities
+
+// Get the basic/default keychain query dictionary for account items.
+// To support searching, the key properties may be partially or fully empty.
+- (NSMutableDictionary *)primaryAttributesForKey:(MSIDCacheKey *)key
+{
+    MSID_TRACE;
+    NSMutableDictionary *query = [self.defaultAccountQuery mutableCopy];
+
+    if (key.account.length > 0)
+    {
+        // Account attribute: <keychainGroup>-<homeAccountId>-<environment>
+        // The access group is added to the account attribute so it's part of the keychain item's primary key.
+        query[(id)kSecAttrAccount] = [NSString stringWithFormat:@"%@-%@", self.keychainGroup, key.account];
+    }
+    if (key.service.length > 0)
+    {
+        // Service attribute: <realm>
+        query[(id)kSecAttrService] = key.service;
+    }
+
+    return query;
+}
+
+// Get the basic/default keychain update dictionary for account items.
+// These are not _primary_ keys, but if they're present in the key object we
+// want to set them when adding/updating.
+- (NSMutableDictionary *)secondaryAttributesForKey:(MSIDCacheKey *)key
+{
+    MSID_TRACE;
+    NSMutableDictionary *query = [NSMutableDictionary new];
+
+    if (key.generic.length > 0)
+    {
+        // Generic attribute: <username>
+        query[(id)kSecAttrGeneric] = key.generic;
+    }
+    if (key.type != nil && [key.type isNotEqualTo:@(1000)])
```

Comment:
shouldn't this be handled by the key? (not being 1000)?
Seems like it makes this component know too much about what values are inside the type

Reply:
I was able to eliminate this reference to 1000 (tracked down calling code in the cpp codebase that was using a key where it should have been using a query object)